exampleCourse/questions/template/number-input/multiple-answer-inputs/server.py

```diff
@@ -0,0 +1,33 @@
+import itertools
+import math
+import random
+
+import sympy
+
+
+def generate(data):
+    # Select 2 random coefficients between -15 and 15, excluding 0.
+    a, b = random.sample(sorted(itertools.chain(range(-15, 0), range(1, 16))), 2)
+
+    # We want to ensure that the polynomial has 2 real roots.
+    # To do this, we make sure that the discriminant is positive.
+    # Solving `b**2 - 4*a*c > 0` for `c`, we get `c < b**2 / (4 * a)`.
+    # Note that the sign of `a` determines the direction of the inequality.
+    c_bound = b**2 / (4 * a)
+    if a > 0:
+        c_bound = math.floor(c_bound)
+        c = random.randint(min(-15, c_bound - 15), c_bound)
+    else:
+        c_bound = math.ceil(c_bound)
+        c = random.randint(c_bound, max(15, c_bound + 15))
```

Comment:
When I (finally) figured this out, I started wondering if this is really the best for a template question. Though, maybe it's useful to demonstrate a technique for constraining the possible answers to a question?

Reply:
Yeah, I considered this too, but I'm generally a little hesitant to recommend this as a strategy. In practice it's probably fine; ChatGPT claims that we'd only have to run that loop ~6 times to have a 99% probability of meeting the condition.

Reply:
Yep, this strategy can of course backfire if the constraints are too restrictive, but I think it's nonetheless a super helpful way to implement things in general. It's typically waaaaaay easier than trying to carefully construct known-good parameters from the start. Even for this super simple quadratic equation case, it's surprisingly tricky to get the logic correct to ensure real roots, and even with your well-designed code we will still end up with some large values of `c` in some cases (not horribly large, but we can get `c = 56` if `a = 1` and `b = 15`).

I've actually wondered in the past about adding explicit support for this pattern by adding a `validate()` function to `server.py`. This would allow PL to do the iteration itself and issue a warning if more than N iterations were required to get a valid set of parameters on average, where N could depend on the CPU cost of a single iteration.

Several years ago Tim and I started discussing a set of common "design patterns" for questions, and iterate-until-constratins-satisfied is a really common one. I'd like to have an explicit list of patterns written down somewhere, with examples, but maybe template questions are also a good way to show these off (also to the LLM). Other common ones are:
* Several correct possible answers, test against all of them. This question could also be an example of this if we ask for the roots in either order, so that (a,b) and (b,a) are both correct answers.
* Reverse generation: generate the correct answer first, then compute the parameters. If you wanted integer roots in the current question then that would be the easiest way to implement it.
* Test for correctness: there might be many correct answers and it's easy to test whether an answer is correct. For example "What is a 2x2 matrix that transforms (1,2) to (-3,7)?".

Reply:
Getting well outside the scope of this PR here, but... I'm hesitant to add support for `validate()` since this is pretty trivial to do with a function + `while` loop, as you point out.

I'm 100% in favor of documenting common strategies like this! Docs + compelling examples would be great, and this is a good start to that.

> If you wanted integer roots in the current question then that would be the easiest way to implement it.

This is actually how I started off writing this question, but I felt that it'd be better to show off support for `<pl-number-input>`, as opposed to `<pl-integer-input>`.

I'll make the switch to generating random parameters in a loop for this question.

Reply:
Now using loop-based parameter generation as of ab7449f.